ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDFConvertCharset.java

```diff
@@ -0,0 +1,138 @@
+import org.apache.hadoop.hive.serde2.objectinspector.ConstantObjectInspector;
+import org.apache.hadoop.hive.serde2.objectinspector.ObjectInspector;
+import org.apache.hadoop.hive.serde2.objectinspector.ObjectInspector.Category;
+import org.apache.hadoop.hive.serde2.objectinspector.PrimitiveObjectInspector;
+import org.apache.hadoop.hive.serde2.objectinspector.primitive.PrimitiveObjectInspectorFactory;
+import org.apache.hadoop.hive.serde2.objectinspector.primitive.PrimitiveObjectInspectorUtils;
+import org.apache.hadoop.hive.serde2.objectinspector.primitive.PrimitiveObjectInspectorUtils.PrimitiveGrouping;
+
+@Description(name = "convertCharset", value = "_FUNC_(str, str, str) - Converts the first argument from the second argument character set to the third argument character set", extended =
+    "Possible options for the character set are 'US-ASCII', 'ISO-8859-1',\n"
+        + "'UTF-8', 'UTF-16BE', 'UTF-16LE', and 'UTF-16'. If either argument\n"
+        + "is null, the result will also be null") public class GenericUDFConvertCharset extends GenericUDF {
+  private transient CharsetEncoder encoder = null;
+  private transient CharsetDecoder decoder = null;
+  private transient PrimitiveObjectInspector stringOI = null;
+  private transient PrimitiveObjectInspector fromCharsetOI = null;
+  private transient PrimitiveObjectInspector toCharsetOI = null;
+
+  @Override public ObjectInspector initialize(ObjectInspector[] arguments) throws UDFArgumentException {
+    if (arguments.length != 3) {
+      throw new UDFArgumentLengthException("ConvertCharset() requires exactly three arguments");
+    }
+
+    if (arguments[0].getCategory() != ObjectInspector.Category.PRIMITIVE
+        || PrimitiveObjectInspectorUtils.PrimitiveGrouping.STRING_GROUP
+        != PrimitiveObjectInspectorUtils.getPrimitiveGrouping(
+        ((PrimitiveObjectInspector) arguments[0]).getPrimitiveCategory())) {
+      throw new UDFArgumentTypeException(0, "The first argument to ConvertCharset() must be a string/varchar");
+    }
+
+    stringOI = (PrimitiveObjectInspector) arguments[0];
+
+    if (arguments[1].getCategory() != ObjectInspector.Category.PRIMITIVE
+        || PrimitiveObjectInspectorUtils.PrimitiveGrouping.STRING_GROUP
+        != PrimitiveObjectInspectorUtils.getPrimitiveGrouping(
+        ((PrimitiveObjectInspector) arguments[1]).getPrimitiveCategory())) {
+      throw new UDFArgumentTypeException(1, "The second argument to ConvertCharset() must be a string/varchar");
+    }
+
+    fromCharsetOI = (PrimitiveObjectInspector) arguments[1];
+
+    if (arguments[2].getCategory() != ObjectInspector.Category.PRIMITIVE
+        || PrimitiveObjectInspectorUtils.PrimitiveGrouping.STRING_GROUP
+        != PrimitiveObjectInspectorUtils.getPrimitiveGrouping(
+        ((PrimitiveObjectInspector) arguments[2]).getPrimitiveCategory())) {
+      throw new UDFArgumentTypeException(2, "The third argument to ConvertCharset() must be a string/varchar");
+    }
```

Comment:
Lots of code duplication here. Is it possible to create a private method with an `int` argument for the `if` block instead?

Reply:
Thanks for your reply, I have modified the code according to the suggestion.
But I ran into some problems in the test, could you give me some advice?
http://ci.hive.apache.org/blue/organizations/jenkins/hive-precommit/detail/PR-3982/5/pipeline/871